custom/src/main/java/org/kiwix/kiwixmobile/custom/download/CustomDownloadFragment.kt

```diff
@@ -56,24 +57,38 @@ class CustomDownloadActivity : BaseActivity() {
   @Inject lateinit var viewModelFactory: ViewModelProvider.Factory
 
   private val compositeDisposable = CompositeDisposable()
+  override fun inject(baseActivity: BaseActivity) {
+    baseActivity.customActivityComponent.inject(this)
+  }
 
-  override fun onCreate(savedInstanceState: Bundle?) {
+  override fun onCreateView(
+    inflater: LayoutInflater,
+    container: ViewGroup?,
+    savedInstanceState: Bundle?
+  ): View? {
     super.onCreate(savedInstanceState)
-    setContentView(R.layout.activity_custom_download)
-    downloadViewModel.state.observe(this, Observer(::render))
+    val root = inflater.inflate(R.layout.fragment_custom_download, container, false)
+    val activity = requireActivity() as CoreMainActivity
+    downloadViewModel.state.observe(viewLifecycleOwner, Observer(::render))
     compositeDisposable.add(
       downloadViewModel.effects.subscribe(
-        { it.invokeWith(this) },
+        { it.invokeWith(activity) },
         Throwable::printStackTrace
       )
     )
+    return root
+  }
+
+  override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
+    super.onViewCreated(view, savedInstanceState)
     cd_download_button.setOnClickListener { downloadViewModel.actions.offer(ClickedDownload) }
     cd_retry_button.setOnClickListener { downloadViewModel.actions.offer(ClickedRetry) }
   }
 
   override fun onDestroy() {
     super.onDestroy()
     compositeDisposable.clear()
+    activity?.finish()
```

Comment:
Anyway we don't want to do all this weird finishing that is a relic of these views being activities we want to navigate between fragments. Replace the finish and start activities with navigate and pops

Reply:
* I'm unable to make an exit via navController. One of the things i tried out was clearing the backstack and then navigating up, this lead to reader screen and when pressed back on it, exited. This doesn't seem good enough to me.
* Also the fragment transaction didn't do the required, in which i removed the current fragment and commited.

Reply:
Show me the code you used in your attempt

Reply:
* `findNavController().navigate(R.id.customReaderFragment)` this never navigates to reader 
```kotlin 
findNavController().apply{
    popBackStack()   // no matter if i use this or not, nothing different happens(i used it thinking, it'll remove reader from stack and then on calling navigateUp() it'll have no other option but to exit)
     navigateUp()
}
```
, this on pressing back in download screen, navigates to reader and when back pressed from here, exits the app.


*  `fragmentManager.beginTransaction().remove(this).commit()` this also does nothing, when used with `commitNow()`, crashes the app saying: a transaction was already being made by fragment manager

Reply:
@Frans-Lukas would know more about our navigation component, thoughts?

instead of navigateUp you could trigger onBack?

Also we should not directly use the fragmentManager though I can see the purpose for purely investigation into the problem.

It might be better if you push up your code so we can get the wider context

Reply:
`onBack()` was giving unresolved reference

Reply:
You would need to use `activity.onBackPressed()` from a fragment

Reply:
This does the exact same job as done by the second attempt  in the second-last comment of mine. If i clear the backStack and then call `onBackPressed()` it said: `java.lang.IllegalStateException: FragmentManager is already executing transactions`

Reply:
onBackPressed doesn't work and finishing activity like this the best solution for now. I'm unable to eliminate crashes without using it(finish). I read about navController and tried few _new_ experiments but none of 'em work, so for the time being this seems to be the best solution.

Reply:
I think [this answer](https://stackoverflow.com/a/56844148) by Ian Lake supports my implementation

Reply:
We don't want to finish our activity, we only have the 1, that is kind of the point about Single Activity Architecture. We don't want to do something weird. Go to the Side Effect FinishAndStartMain and stop it doing that first of all, then alter the code so we use a nav controller to preform reasonable navigation

Reply:
Why do we want to conserve the activity when the app is closed, just because the docs say so?

Reply:
> then alter the code so we use a nav controller to preform reasonable navigation

Via navController we navigate to fragments only, it doesn't provide us with any method to directly exit out. And I've already tried the navController multiple times differently(that I've mentioned previously) via clearing the backtstack and then navigating up, it just does nothing.

Reply:
>Why do we want to conserve the activity when the app is closed,

I don't understand this sentence. For the purposes of navigation all the responsibility should be delegated to our navigation component. Restarting the activity because we are too lazy to navigate properly is surprising, the more surprising things you find in a code base the higher the "what the fuck"s per minute and that generally makes engineers unhappy, to be cursing their codebase constantly

Reply:
We aren't trying to exit out, what gave you that idea? 
The old system:
Launch App
Be at Main Activity
No custom zim file detected so go to CustomDownload Activity (also finish Main so users can't go back from Download)
Download a Zim File
Download Completes,  go to Main (also finish activity so users can't go back to Download)

We want to keep this behaviour but with Navigation instead of Activities and I am not seeing how this finish call and the lack of changes to FinishAndStartMain achieve this

Reply:
> We aren't trying to exit out, what gave you that idea?

Then what else one expects on landing at download screen and not interested in downloading the file and therefore presses Back button, probably simply the person wants to exit out. And the current changes in PR do produce the same UX as mentioned the procedure described above by you. 

The thing is we have now download as a fragment so we can't finish it like we did for the activity and via navController I can navigate only to the fragments, so now if I clear the backstack of Download(which comprises of fragments only) and then call navigate up, it does nothing, the back press doesn't produce any effect, this is the solo reason why I finshed the mainActivity, that's it. I am in complete harmony with the no. of wtf's ideology that you mentioned above, so I'll give it a loook again

Reply:
A back press should navigate backwards in the navigation stack, if the stack is empty then the app should exit, this functionality should not be implemented here, it should be implicit in the system otherwise we would have to implement it everywhere. You can see the back press handling in CoreMainActivity

```
  override fun onBackPressed() {
    if (navigationDrawerIsOpen()) {
      closeNavigationDrawer()
      return
    }
    if (activeFragments().filterIsInstance<FragmentActivityExtensions>().isEmpty()) {
      return super.onBackPressed()
    }
    activeFragments().filterIsInstance<FragmentActivityExtensions>().forEach {
      if (it.onBackPressed(this) == FragmentActivityExtensions.Super.ShouldCall) {
        super.onBackPressed()
      }
    }
  }
```

and various implementations of FragmentActivityExtensions where we explicitly handle `onBackPressed` in fragments, potentially all you need to do is have CustomDownloadFragment implement FragmentActivityExtensions

Reply:
Back press now works as expected in the scenarios accordingly.